src/expression/comparison/ge.ts

```diff
@@ -31,9 +31,13 @@ export class GreaterThanOrEqual extends Comparison {
    * {@link Comparison.comparison}
    */
   comparison (left: Result, right: Result): boolean {
-    if (isNumber(left) && isNumber(right)) {
+    const operand = (value: Result) =>
+      this.strict ? value : this.toNumber(value)
+
+    if (isNumber(operand(left)) && isNumber(operand(right))) {
       return (left as number) >= (right as number)
```

Comment:
This will work as expected - unless I am missing something.
`isNumber(operand(left))`
 - if typed and evaluated as number, good 
 - but than here `(left as number)` you are still using the original variable - i.e. not typed to number,

Reply:
I guess it "works as expected" because it's tricking typescript. For regular javascript `2 > '1'` works, so the tests pass, even if `toNumber` is not used for the actual comparison.

Reply:
TIL: `'2.3' > '2.2'` is valid :D

Reply:
> TIL: `'2.3' > '2.2'` is valid :D

I was also shocked to see that this works, '2010-01-01' < '2009-02-03' as well

Reply:
Okay, so I would actually propose that it in `strict` mode (default):
 - `ge, gt, le, lt` would evaluate to false, if any of the operands is **NOT** a numbers
 - i.e. the same behavior as `eq, ne` (achieved by the === value and type check natively by JS)
- This is avoid these unexpected "side effects" from the DSL perspective.

@RyanNieuwoudt @lilyvanoekel @theslyone thoughts?

Reply:
This will be a breaking change because default mode is strict but will cause `'2010-01-01' > '2009-02-03' = false` and `2 > '1' = false` when js by default allows this to pass

Reply:
Understand, but `JS` implementation of the illogical DSL. If 3rd party will implement our questionnaire it other languages, that would not expect `2 > '1'` to evaluate as true.

And, do we actually some conditions which will break?

Reply:
> Understand, but this is just JS implementation of the illogical DSL. If 3rd party will implement our questionnaire in other languages, than they would not expect 2 > '1' to evaluate as true.

Do you know of any third party usage of this particular js library currently? I'm saying this will break their implementation and so we need to capture this in our breaking changes.

> And, do we actually some conditions which will break?

We cannot be certain at this point the implication of this on briza-api although we plan to use it with `strict=false` so we should be good

Reply:
I think without @theslyone changes it checks if things are a number so `'2010-01-01' > '2009-02-03'` will return false?

I'm not so sure about other languages and what happens to `'2010-01-01' > '2009-02-03'`. This seems like a normal string compare, no?

Agree on `2 > '1'`

Reply:
https://stackoverflow.com/questions/8406703/c-compare-to-string-dates

Reply:
> We cannot be certain at this point the implication of this on briza-api although we plan to use it with strict=false so we should be good

This should be testable. "Simply" traversing trough all questions condition props - @RyanNieuwoudt you've just written something like this for detection of conflicts in conditions - any suggestions?

Reply:
> Do you know of any third party usage of this particular js library currently? I'm saying this will break their implementation and so we need to capture this in our breaking changes.

Not sure, will double check. Though, it would "break" only if we have some issues in the conditions which is TBD, and we can fix them when detected.

Reply:
maybe we should not have strict mode. Instead just force both operands to be of the same type?

We should probably also define what we want to happen with `null` and `undefined`

Reply:
> '2010-01-01' > '2009-02-03'. This seems like a normal string compare, no? Yes, string compare (though even that is not normal in other major languages) but it evokes the feeling that it is actually comparing dates.

Reply:
> maybe we should not have strict mode. Instead just force both operands to be of the same type?

This will bring us back to our original issues below wherein we'll be relying solely on js to ensure `'10' > '0' = true` 
```ts
[
  ">",
  "$Location1BuildingLimit",
  "0"
]```

Reply:
We could specifically check for the types number or string. I'm sure every language has some kind of string compare function.

Reply:
> [">", "$Location1BuildingLimit", "0"]

we should fix these to `[">", "$Location1BuildingLimit", 0]`

Reply:
> we should fix these to [">", "$Location1BuildingLimit", 0]

`$Location1BuildingLimit` is a `currency` question type so answer is a string hence the problem.

We want to enable type coercion by setting `strict = false` so that illogical can resolve `[">", "$Location1BuildingLimit", 0]` properly

Reply:
> Do you know of any third party usage of this particular js library currently?

Quick intel from Luiz (not to be taken word by word): "TowerIQ is java and I believe that they were going to implement their own version of the condition evaluator (Rishi might have better details). Quiktract is php and I also believe that they were going to implement their own version. I can check with Nick"

Reply:
> $Location1BuildingLimit is a currency question type so answer is a string hence the problem.

I see. 

Shall we than define/describe that `ge, gt, le, lt` should convert parse-able strings to numbers?
- By definition of the illogical DSL?
- Therefore actually remove the `strict` mode, and rather follow this set standard.
- If 3rd parties will implement illogical DSL, than they would understand what is expected.

Reply:
Another very interesting idea (@undees): 

Explicit optional reference casting: `[">", "$Ref.(Number)", 0]`

Reply:
I'll investigate this and probably close this PR in favour of another to avoid :wrench: :hammer: :fire: :fire:

Reply:
Btw, it may worthwhile to create first a PR which will include only the prettier formatting update - to make the PR super focused on the actual develop feature.

Reply:
I really like where @undees is going with that. Could this include the ability to perform any `map` operation on the data before the comparison takes place?

Allowing for explicit conversions:

1. Guards against the principle of least surprise (for non JavaScript developers in particular).
2. Could be more flexible than just introducing JavaScript's coercion behaviours.
3. May require less code in Illogical itself, deferring to the consuming code for circumstantial logic.

Reply:
> Avoids adding the strictMode option

In this case https://github.com/briza-insurance/illogical/pull/137#discussion_r574627064 ... it should not be needed.